### PR TITLE
ENG 2439 add engine and resource to metrics and checks

### PR DESCRIPTION
**Pull request by likawind**
## Describe your changes and why you are making these changes
This PR adds engine and resource configs for metrics and checks so that the behavior is consistent to `@op()` decorators. We simply moved the relevant parts in `@op()` decorator to a helper function and applies it to both `check()` and `metrics()` implementations.

## Related issue number (if any)
ENG 2439

## Loom demo (if any)
Testing WIP

## Checklist before requesting a review
- [ ] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [ ] I have performed a self-review of my code.
- [ ] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [ ] If this is a new feature, I have added unit tests and integration tests.
- [ ] I have run the integration tests locally and they are passing.
- [ ] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [ ] All features on the UI continue to work correctly.
- [ ] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

### CODE REVIEW

**Comment by likawind** [comment]
@kenxu95 what's the easiest way to test these changes if I don't have an existing engine set up?

**Comment by kenxu95** [comment]
You can try rebasing on https://github.com/aqueducthq/aqueduct/pull/1046 and running the periodic integration tests manually. Otherwise, I'd recommend setting up your own cluster - it's super easy with Terraform.

File: sdk/aqueduct/decorator.py [hunk 1]
```diff
@@ -368,6 +368,62 @@ def _convert_memory_string_to_mbs(memory_str: str) -> int:
     return multiplier * int(memory_scalar_str)
 
 
+def _update_operator_spec_with_engine_and_resource(
```
Comment (kenxu95) [inline sdk/aqueduct/decorator.py:371]: Can we split this into two methods, one for updating `engine` and one for `resources`.

File: sdk/aqueduct/decorator.py [hunk 2]
```diff
@@ -368,6 +368,62 @@ def _convert_memory_string_to_mbs(memory_str: str) -> int:
...
+    spec: OperatorSpec,
+    engine: Optional[str] = None,
+    resources: Optional[Dict[str, Any]] = None,
```
Comment (kenxu95) [inline sdk/aqueduct/decorator.py:374]: I wonder if our resource keys should be enums?

File: src/golang/lib/models/shared/operator/operator.go [hunk 3]
```diff
@@ -53,7 +53,8 @@ type specUnion struct {
 	Param        *param.Param                `json:"param,omitempty"`
 	SystemMetric *system_metric.SystemMetric `json:"system_metric,omitempty"`
 
-	// This can currently only be set for function operators.
+	// This can currently only be set for operators that has function,
```
Comment (kenxu95) [inline src/golang/lib/models/shared/operator/operator.go:56]: maybe "for function operators, including metrics and checks."

**Review by kenxu95 — COMMENTED** [review]
Cool! One small refactor request and we should be good to go

**Comment by vsreekanti** [comment]
Quick question on this — is it possible to share the logic for these flags in some way such that we automatically inherit changes between `@op`, `@metric`, and` @check`?

**Comment by likawind** [comment]
@vsreekanti good point, it requires us to refactor function decorator to class decorator (and use the inheritance pattern) so I'm putting a linear task for this:

https://linear.app/aqueducthq/issue/ENG-2575/allow-metric-and-check-easily-inherit-op-decorator-inputs

**Review by cw75 — APPROVED** [review]

**Comment by likawind** [comment]
Mypy failures are known to happen on main, merging giving it's passing locally.

**Pull request merged**